File: Parcial2024.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <queue>
#include <stack>
#include <unordered_map>

using namespace std;

struct Libro {
    string titulo;
    string autor;
    int anoEdicion;
    string editorial;
    string isbn;
    int paginas;
    bool disponible;
    Libro* siguiente;
};

struct Lector {
    string nombre;
    int dni;
    string libroSolicitado;
    Lector* siguiente;
};

Libro* listaLibros = nullptr;
queue<Lector*> colaSolicitudes;
stack<string> historial;
unordered_map<string, string> librosSolicitados;

void agregarLibro(Libro*& lista, const string& titulo, const string& autor, int ano, const string& editorial, const string& isbn, int paginas, bool disponible);
void cargarLibros(Libro*& lista);
void guardarLibros(Libro* lista);
void ordenarLibrosPorTitulo(Libro*& lista);
void mostrarHistorial();
void vaciarBiblioteca(Libro*& lista);
Libro* buscarLibroPorTitulo(Libro* lista, const string& titulo);
void solicitarLibro(queue<Lector*>& cola, const string& nombre, int dni, const string& libroSolicitado);
void devolverLibro(queue<Lector*>& cola, const string& libroDevuelto);
void guardarSolicitud(Lector* lector); // Prototipo de la función

void agregarLibro(Libro*& lista, const string& titulo, const string& autor, int ano, const string& editorial, const string& isbn, int paginas, bool disponible) {
    Libro* nuevoLibro = new Libro{titulo, autor, ano, editorial, isbn, paginas, disponible, nullptr};
    if (!lista) {
        lista = nuevoLibro;
    } else {
        Libro* actual = lista;
        while (actual->siguiente) {
            actual = actual->siguiente;
        }
        actual->siguiente = nuevoLibro;
    }
    historial.push("Libro agregado: " + titulo);
}

void cargarLibros(Libro*& lista) {
    ifstream archivo("biblioteca.txt");
    if (archivo.is_open()) {
        string titulo, autor, editorial, isbn;
        int ano, paginas;
        while (getline(archivo, titulo, ',') &&
               getline(archivo, autor, ',') &&
               archivo >> ano &&
               archivo.ignore() &&
               getline(archivo, editorial, ',') &&
               getline(archivo, isbn, ',') &&
               archivo >> paginas &&
               archivo.ignore()) {
            agregarLibro(lista, titulo, autor, ano, editorial, isbn, paginas, true);
        }
        archivo.close();
        historial.push("Datos de libros cargados");
    } else {
        cout << "No se pudo abrir el archivo para cargar." << endl;
    }
}

void guardarLibros(Libro* lista) {
    ofstream archivo("biblioteca.txt");
    if (archivo.is_open()) {
        for (Libro* actual = lista; actual != nullptr; actual = actual->siguiente) {
            if (actual->disponible) {
                archivo << actual->titulo << "," << actual->autor << "," << actual->anoEdicion << ","
                        << actual->editorial << "," << actual->isbn << "," << actual->paginas << endl;
            }
        }
        archivo.close();
    } else {
        cout << "No se pudo abrir el archivo para guardar." << endl;
    }
}

void ordenarLibrosPorTitulo(Libro*& lista) {
    bool ordenado = false;
    while (!ordenado) {
        ordenado = true;
        Libro* actual = lista;
        Libro* prev = nullptr;
        while (actual && actual->siguiente) {
            if (actual->titulo > actual->siguiente->titulo) {
                if (prev) {
                    prev->siguiente = actual->siguiente;
                } else {
                    lista = actual->siguiente;
                }
                Libro* temp = actual->siguiente->siguiente;
                actual->siguiente->siguiente = actual;
                actual->siguiente = temp;
                ordenado = false;
                continue;
            }
            prev = actual;
            actual = actual->siguiente;
        }
    }
    historial.push("Libros ordenados por título");
}

void mostrarHistorial() {
    cout << "Historial de operaciones:\n";
    stack<string> aux = historial;
    while (!aux.empty()) {
        cout << aux.top() << endl;
        aux.pop();
    }
}

void vaciarBiblioteca(Libro*& lista) {
    while (lista) {
        Libro* temp = lista;
        lista = lista->siguiente;
        delete temp;
    }
    ofstream archivo("biblioteca.txt", ios::trunc);
    archivo.close();
    historial.push("Biblioteca vaciada");
    cout << "La biblioteca ha sido vaciada." << endl;
}

void vaciarSolicitudes() {
    ofstream archivo("solicitudes.txt", ios::trunc); // Abre el archivo y elimina el contenido existente
    if (archivo.is_open()) {
        cout << "El archivo de solicitudes ha sido vaciado." << endl;
        archivo.close();
    } else {
        cout << "No se pudo abrir el archivo para vaciar." << endl;
    }
}


Libro* buscarLibroPorTitulo(Libro* lista, const string& titulo) {
    Libro* actual = lista;
    while (actual) {
        if (actual->titulo == titulo) {
            return actual;
        }
        actual = actual->siguiente;
    }
    return nullptr;
}

void solicitarLibro(queue<Lector*>& cola, const string& nombre, int dni, const string& libroSolicitado) {
    Libro* libro = buscarLibroPorTitulo(listaLibros, libroSolicitado);
    if (!libro || !libro->disponible) {
        cout << "No encontramos el libro que solicitaste o ya está solicitado." << endl;
        return;
    }
    libro->disponible = false;
    Lector* nuevoLector = new Lector{nombre, dni, libroSolicitado, nullptr};
    cola.push(nuevoLector);
    historial.push("Solicitud libro: " + nombre + " para " + libroSolicitado);
    guardarSolicitud(nuevoLector);  // Llama a guardar en el archivo de solicitudes
    guardarLibros(listaLibros);
}

void devolverLibro(queue<Lector*>& cola, const string& libroDevuelto) {
    Libro* libro = buscarLibroPorTitulo(listaLibros, libroDevuelto);
    if (libro && !libro->disponible) {
        libro->disponible = true;
        cout << "Libro '" << libroDevuelto << "' ha sido devuelto y está disponible para ser solicitado nuevamente." << endl;
        guardarLibros(listaLibros);
    } else {
        cout << "Este libro no estaba marcado como solicitado o ya fue devuelto." << endl;
    }
}

void guardarSolicitud(Lector* lector) {
    ofstream archivo("solicitudes.txt", ios::app); // Abre el archivo en modo de agregar
    if (archivo.is_open()) {
        archivo << lector->nombre << "," << lector->dni << "," << lector->libroSolicitado << "\n";
        archivo.close();
    } else {
        cout << "No se pudo abrir el archivo para guardar solicitudes." << endl;
    }
}

int main() {
    int opcion;
    cargarLibros(listaLibros); // Cargar datos desde archivo al iniciar

    do {
        cout << "\nSistema de Gestión de Biblioteca\n";
        cout << "1. Agregar libro\n";
        cout << "2. Ordenar libros por título\n";
        cout << "3. Buscar libro por título\n";
        cout << "4. Solicitar libro\n";
        cout << "5. Devolver libro\n";
        cout << "6. Guardar y salir\n";
        cout << "7. Ver historial de operaciones\n";
        cout << "8. Vaciar biblioteca\n";
        cout << "9. Vaciar archivo de solicitudes\n";
        cout << "0. Salir\n";
        cout << "Opción: ";
        cin >> opcion;
        cin.ignore();

        switch (opcion) {
            case 1: {
                string titulo, autor, editorial, isbn;
                int ano, paginas;
                cout << "Título: "; getline(cin, titulo);
                cout << "Autor: "; getline(cin, autor);

                // Validación de entrada del año de edición
                while (true) {
                    cout << "Año de edición: ";
                    if (!(cin >> ano)) {
                        cin.clear(); // Limpia el estado de error de cin
                        cin.ignore(1000, '\n'); // Ignora hasta 1000 caracteres o hasta el próximo '\n'
                        cout << "Entrada no válida. Ingrese un número para el año de edición.\n";
                    } else {
                        cin.ignore(); // Limpia cualquier sobrante en el buffer
                        break;
                    }
                }

                cout << "Editorial: "; getline(cin, editorial);
                cout << "ISBN: "; getline(cin, isbn);
                cout << "Páginas: "; cin >> paginas; cin.ignore();
                agregarLibro(listaLibros, titulo, autor, ano, editorial, isbn, paginas, true);
                guardarLibros(listaLibros);
                break;
            }
            case 2:
                ordenarLibrosPorTitulo(listaLibros);
                cout << "Libros ordenados por título.\n";
                break;
            case 3: {
                string titulo;
                cout << "Ingrese el título del libro a buscar: "; getline(cin, titulo);
                Libro* encontrado = buscarLibroPorTitulo(listaLibros, titulo);
                if (encontrado) {
                    cout << "Libro encontrado: " << encontrado->titulo << " de " << encontrado->autor << endl;
                } else {
                    cout << "No se encontró el libro o no está disponible.\n";
                }
                break;
            }
            case 4: {
                string nombre, titulo;
                int dni;
                cout << "Nombre del lector: "; getline(cin, nombre);
                cout << "DNI del lector: "; cin >> dni; cin.ignore();
                cout << "Libro solicitado: "; getline(cin, titulo);
                solicitarLibro(colaSolicitudes, nombre, dni, titulo);
                break;
            }
            case 5: {
                string libroDevuelto;
                cout << "Libro devuelto: "; getline(cin, libroDevuelto);
                devolverLibro(colaSolicitudes, libroDevuelto);
                break;
            }
            case 6:
                guardarLibros(listaLibros);
                cout << "Datos guardados. Saliendo...\n";
                break;
            case 7:
                mostrarHistorial();
                break;
            case 8:
                vaciarBiblioteca(listaLibros);
                break;
            case 9:
                 vaciarSolicitudes();
                break;

            case 0:
                cout << "Saliendo...\n";
                break;
            default:
                cout << "Opción no válida.\n";
        }
    } while (opcion != 0);

    return 0;
}

File: Parcial2024.py
import csv
from collections import deque

class Libro:
    def __init__(self, titulo, autor, ano_edicion, editorial, isbn, paginas, disponible=True):
        self.titulo = titulo
        self.autor = autor
        self.ano_edicion = ano_edicion
        self.editorial = editorial
        self.isbn = isbn
        self.paginas = paginas
        self.disponible = disponible

class Lector:
    def __init__(self, nombre, dni, libro_solicitado):
        self.nombre = nombre
        self.dni = dni
        self.libro_solicitado = libro_solicitado

# Variables globales
lista_libros = []
cola_solicitudes = deque()
historial = []

# Funciones
def agregar_libro(titulo, autor, ano, editorial, isbn, paginas):
    nuevo_libro = Libro(titulo, autor, ano, editorial, isbn, paginas)
    lista_libros.append(nuevo_libro)
    historial.append(f"Libro agregado: {titulo}")
    guardar_libros()

def cargar_libros():
    try:
        with open("bibliotecapython.csv", newline='', encoding="utf-8") as archivo:
            lector_csv = csv.reader(archivo)
            for fila in lector_csv:
                titulo, autor, ano, editorial, isbn, paginas = fila
                agregar_libro(titulo, autor, int(ano), editorial, isbn, int(paginas))
        print("Datos de libros cargados")
    except FileNotFoundError:
        print("No se pudo abrir el archivo para cargar.")

def guardar_libros():
    with open("bibliotecapython.csv", mode="w", newline='', encoding="utf-8") as archivo:
        escritor_csv = csv.writer(archivo)
        for libro in lista_libros:
            if libro.disponible:
                escritor_csv.writerow([libro.titulo, libro.autor, libro.ano_edicion, 
                                       libro.editorial, libro.isbn, libro.paginas])

def ordenar_libros_por_titulo():
    lista_libros.sort(key=lambda libro: libro.titulo)
    historial.append("Libros ordenados por titulo")
    guardar_libros()
    print("Libros ordenados por título y guardados en bibliotecapython.csv.")

def mostrar_historial():
    print("Historial de operaciones:")
    for operacion in historial:
        print(operacion)

def buscar_libro_por_titulo(titulo):
    for libro in lista_libros:
        if libro.titulo == titulo:
            return libro
    return None

def solicitar_libro(nombre, dni, titulo_libro):
    libro = buscar_libro_por_titulo(titulo_libro)
    if not libro or not libro.disponible:
        print("No encontramos el libro que solicitaste o ya está solicitado.")
        return

    libro.disponible = False
    nuevo_lector = Lector(nombre, dni, titulo_libro)
    cola_solicitudes.append(nuevo_lector)
    historial.append(f"Solicitud libro: {nombre} para {titulo_libro}")
    guardar_solicitud(nuevo_lector)
    guardar_libros()

def devolver_libro(titulo_libro):
    libro = buscar_libro_por_titulo(titulo_libro)
    if libro and not libro.disponible:
        libro.disponible = True
        print(f"Libro '{titulo_libro}' ha sido devuelto y está disponible para ser solicitado nuevamente.")
        guardar_libros()
    else:
        print("Este libro no estaba marcado como solicitado o ya fue devuelto.")

def vaciar_biblioteca():
    lista_libros.clear()
    with open("bibliotecapython.csv", "w", encoding="utf-8") as archivo:
        pass  # Esto vacía el contenido del archivo
    historial.append("Biblioteca vaciada")
    print("La biblioteca ha sido vaciada.")

def vaciar_solicitudes():
    cola_solicitudes.clear()
    with open("solicitudpython.txt", "w", encoding="utf-8") as archivo:
        pass  # Esto vacía el contenido del archivo
    print("El archivo de solicitudes ha sido vaciado.")

def guardar_solicitud(lector):
    with open("solicitudpython.txt", "a", encoding="utf-8") as archivo:
        archivo.write(f"{lector.nombre},{lector.dni},{lector.libro_solicitado}\n")

# Menú principal
def main():
    cargar_libros()

    while True:
        print("\nSistema de Gestión de Biblioteca")
        print("1. Agregar libro")
        print("2. Ordenar libros por título")
        print("3. Buscar libro por título")
        print("4. Solicitar libro")
        print("5. Devolver libro")
        print("6. Guardar y salir")
        print("7. Ver historial de operaciones")
        print("8. Vaciar biblioteca")
        print("9. Vaciar archivo de solicitudes")
        print("0. Salir")
        opcion = input("Opción: ")

        if opcion == "1":
            titulo = input("Título: ")
            autor = input("Autor: ")

            # Validación para el año de edición
            while True:
                try:
                    ano = int(input("Año de edición: "))
                    break
                except ValueError:
                    print("Ingrese un número para el año de edición.")

            editorial = input("Editorial: ")
            isbn = input("ISBN: ")

            # Validación para la cantidad de páginas
            while True:
                try:
                    paginas = int(input("Páginas: "))
                    break
                except ValueError:
                    print("Ingrese un número para la cantidad de páginas.")

            agregar_libro(titulo, autor, ano, editorial, isbn, paginas)

        elif opcion == "2":
            ordenar_libros_por_titulo()

        elif opcion == "3":
            titulo = input("Ingrese el título del libro a buscar: ")
            libro = buscar_libro_por_titulo(titulo)
            if libro:
                print(f"Libro encontrado: {libro.titulo} de {libro.autor}")
            else:
                print("No se encontró el libro o no está disponible.")

        elif opcion == "4":
            nombre = input("Nombre del lector: ")
            dni = input("DNI del lector: ")
            titulo_libro = input("Libro solicitado: ")
            solicitar_libro(nombre, dni, titulo_libro)

        elif opcion == "5":
            titulo_libro = input("Libro devuelto: ")
            devolver_libro(titulo_libro)

        elif opcion == "6":
            guardar_libros()
            print("Datos guardados. Saliendo...")
            break

        elif opcion == "7":
            mostrar_historial()

        elif opcion == "8":
            vaciar_biblioteca()

        elif opcion == "9":
            vaciar_solicitudes()

        elif opcion == "0":
            print("Saliendo...")
            break

        else:
            print("Opción no válida.")

if __name__ == "__main__":
    main()




